sql/expression/function/math.go

```diff
@@ -105,4 +111,236 @@ func (r *Rand) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
 	}
 
 	return rand.New(rand.NewSource(seed)).Float64(), nil
-}
\ No newline at end of file
+}
+
+var UintRegex, _ = regexp.Compile("^[1-9][0-9]*$*")
+var IntRegex, _ = regexp.Compile("^-?[1-9][0-9]*$")
+
+type UnaryMathFloatFuncLogic interface {
+	EvalFloat(float64) (interface{}, error)
+}
+
+type UnaryMathFuncLogic interface {
+	UnaryMathFloatFuncLogic
+	EvalUint(uint64) (interface{}, error)
+	EvalInt(int64) (interface{}, error)
+	EvalDecimal(decimal.Decimal) (interface{}, error)
+}
+
+type UnaryMathFloatFuncWrapper struct {
+	FlLogic UnaryMathFloatFuncLogic
+}
+
+func WrapUnaryMathFloatFuncLogic(logic UnaryMathFloatFuncLogic) UnaryMathFuncLogic {
+	return UnaryMathFloatFuncWrapper{logic}
+}
+
+func (wr UnaryMathFloatFuncWrapper) EvalFloat(n float64) (interface{}, error) {
+	return wr.FlLogic.EvalFloat(n)
+}
+
+func (wr UnaryMathFloatFuncWrapper) EvalUint(n uint64) (interface{}, error) {
+	return wr.FlLogic.EvalFloat(float64(n))
+}
+func (wr UnaryMathFloatFuncWrapper) EvalInt(n int64) (interface{}, error) {
+	return wr.FlLogic.EvalFloat(float64(n))
+}
+
+func (wr UnaryMathFloatFuncWrapper) EvalDecimal(dec decimal.Decimal) (interface{}, error) {
+	n, _ := dec.Float64()
+	return wr.FlLogic.EvalFloat(n)
+}
+
+type UnaryMathFunc struct {
+	expression.UnaryExpression
+	Name  string
+	Logic UnaryMathFuncLogic
+}
+
+func NewUnaryMathFunc(name string, logic UnaryMathFuncLogic) func(e sql.Expression) sql.Expression {
+	return func(e sql.Expression) sql.Expression {
+		return &UnaryMathFunc{expression.UnaryExpression{Child: e}, name, logic}
+	}
+}
+
+// Eval implements the Expression interface.
+func (mf *UnaryMathFunc) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	val, err := mf.Child.Eval(ctx, row)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if val == nil {
+		return nil, nil
+	}
+
+	// Fucking Golang
+	switch x := val.(type) {
+	case uint8:
+		return mf.Logic.EvalUint(uint64(x))
+	case uint16:
+		return mf.Logic.EvalUint(uint64(x))
+	case uint32:
+		return mf.Logic.EvalUint(uint64(x))
+	case uint64:
+		return mf.Logic.EvalUint(x)
+	case uint:
+		return mf.Logic.EvalUint(uint64(x))
+	case int8:
+		return mf.Logic.EvalInt(int64(x))
+	case int16:
+		return mf.Logic.EvalInt(int64(x))
+	case int32:
+		return mf.Logic.EvalInt(int64(x))
+	case int64:
+		return mf.Logic.EvalInt(x)
+	case int:
+		return mf.Logic.EvalInt(int64(x))
+	case float64:
+		return mf.Logic.EvalFloat(x)
+	case float32:
+		return mf.Logic.EvalFloat(float64(x))
+	case decimal.Decimal:
+		return mf.Logic.EvalDecimal(x)
+	case string:
```

Comment:
For non-numeric types, you should just call Float.Convert, not use your own logic

Reply:
Actually consider just doing Float.Convert for everything. You're casting everything to floats anyway.

Reply:
What do you mean I'm casting to floats anyway?

Reply:
The trig functions cast everything to floats... but not all math functions will.

Reply:
I missed that detail. I think this is fine as long as you get rid of the custom string parsing and use Float.Convert instead. The other way to handle it, if you really wanted to support string -> int conversion, would be to have the math function interface declare a TargetType, then call type.Convert with that.